src/db/spanner/batch.rs

```diff
@@ -335,23 +370,63 @@ pub async fn do_append_async(
     .into_iter()
     .map(|(name, field_type)| struct_type_field(name, field_type))
     .collect();
-    struct_type.set_fields(RepeatedField::from_vec(fields));
-    array_type.set_struct_type(struct_type);
-    param_type.set_array_element_type(array_type);
-
-    let mut sqlparams = HashMap::new();
-    sqlparams.insert("values".to_owned(), values);
-    let mut sqlparam_types = HashMap::new();
-    sqlparam_types.insert("values".to_owned(), param_type);
-    db.sql(
-        "INSERT INTO batch_bsos (fxa_uid, fxa_kid, collection_id, batch_id, batch_bso_id,
-                                 sortindex, payload, ttl)
-         SELECT * FROM UNNEST(@values)",
-    )?
-    .params(sqlparams)
-    .param_types(sqlparam_types)
-    .execute_dml_async(&db.conn)
-    .await?;
+
+    if !insert.is_empty() {
+        let mut list_values = ListValue::new();
+        list_values.set_values(RepeatedField::from_vec(insert));
+        let mut values = Value::new();
+        values.set_list_value(list_values);
+
+        // values' type is an ARRAY of STRUCTs
+        let mut param_type = Type::new();
+        param_type.set_code(TypeCode::ARRAY);
+        let mut array_type = Type::new();
+        array_type.set_code(TypeCode::STRUCT);
+
+        // STRUCT requires definition of all its field types
+        let mut struct_type = StructType::new();
+        struct_type.set_fields(RepeatedField::from_vec(fields));
+        array_type.set_struct_type(struct_type);
+        param_type.set_array_element_type(array_type);
+
+        let mut sqlparams = HashMap::new();
+        sqlparams.insert("values".to_owned(), values);
+        let mut sqlparam_types = HashMap::new();
+        sqlparam_types.insert("values".to_owned(), param_type);
+        db.sql(
+            "INSERT INTO batch_bsos (fxa_uid, fxa_kid, collection_id, batch_id, batch_bso_id,
+                                    sortindex, payload, ttl)
+            SELECT * FROM UNNEST(@values)",
+        )?
+        .params(sqlparams)
+        .param_types(sqlparam_types)
+        .execute_dml_async(&db.conn)
+        .await?;
+    }
+
+    // assuming that "update" is rarer than an insert, we can try using the standard API for that.
+    if !update.is_empty() {
+        for val in update {
+            db.sql(
+                "UPDATE batch_bsos SET sortindex=@sortindex, payload=@payload, ttl=@ttl
```

Comment:
I think this works slightly different than the Python INSERT ON CONFLICT UPDATE version. That version only issues UPDATES for batch columns that are NOT NULL.

If fixing this be painful we could ping @rfk about the behavior first.

Reply:
Hrm, yeah, I guess you can imagine a weird-but-technically-allowed client sending two partial updates in a batch, like `{"id": "foo", "payload": "bar"}` and then later `{"id": "foo", "ttl": 27}` in the same batch. The correct semantics would be to apply them as two partial updates, resulting in `{"id": "foo", "payload": "bar", "ttl": 27}`, but it would be even better if we could just convince ourselves that current clients don't do that, and then forbid it.